blockstack_client/backend/drivers/ipfs.py

```diff
@@ -0,0 +1,410 @@
+    We create it, and ignore the error that throws if it exists already.
+    '''
+    api = ipfsapi.connect('127.0.0.1', 5001)
+    try:
+      r = api._client.request('/key/gen', (key,), decoder='json', opts={'type':'rsa','size':'2048'})
+    except ipfsapi.exceptions.ErrorResponse:
+      # An exception is thrown when the key already exists, we ignore it
+      pass
+
+def write_chunk( chunk_path, chunk_buf, is_mutable ):
+    """
+    Write a chunk of data to IPFS.
+    
+    Return True on success 
+    Return False on error, and log an exception
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    rc = True
+    begin = None
+    end = None
+    size = None
+    try:
+        if IPFS_COMPRESS:            
+            compressed_data = compress_chunk( chunk_buf )
+        else:
+            compressed_data = chunk_buf
+
+        size = len(compressed_data)
+
+        begin = time.time()
+        h = api.add_str( compressed_data, decoder='json' )
+        if is_mutable:
+          if os.environ.get("BLOCKSTACK_TEST") == "1":
+            log.debug("[DEBUG] ipfs.publish_name %s: %s" % (chunk_path, chunk_buf))
+          r = api.name_publish( h, opts = {'key': 'mutable-{}'.format(chunk_path.replace( "/", r"\x2f" ))})
+          if os.environ.get("BLOCKSTACK_TEST") == "1":
+            log.debug("[DEBUG] ipfs.publish_name returned " % (r))
+        end = time.time()
+        
+    except Exception, e:
+        log.error("Failed to write '%s'" % chunk_path)
+        log.exception(e)
+        rc = False
+
+    if os.environ.get("BLOCKSTACK_TEST") == "1" and rc:
+        log.debug("[BENCHMARK] s3.write_chunk %s: %s" % (size, end - begin))
+    return rc
+
+#-------------------------
+def read_chunk( chunk_path, is_mutable ):
+    """
+    Get a chunk of data from IPFS.
+    
+    Return the data on success
+    Return None on error, and log an exception.
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    data = None
+    begin = None
+    end = None
+    size = None
+    try:
+        begin = time.time()
+        if is_mutable:
+          r = api.name_resolve( chunk_path )
+          url = r['Path']
+        else:
+          url = chunk_path
+        compressed_data = api.cat(url)
+        end = time.time()
+        size = len(compressed_data)
+
+        try:
+            data = decompress_chunk( compressed_data )
+        except:
+            data = compressed_data
+        
+    except Exception, e:
+        log.error("Failed to read '%s'" % chunk_path)
+        log.exception(e)
+        
+    if os.environ.get("BLOCKSTACK_TEST") == "1":
+        log.debug("[BENCHMARK] s3.read_chunk %s: %s" % (size, end - begin))
+    return data
+
+#-------------------------
+def delete_chunk( chunk_path, is_mutable ):
+    """
+    Delete a chunk of data from IPFS.
+    
+    Return True on success 
+    Return False on error.
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
```

Comment:
you probably don't want to make a new connection for each command, i would make one with the instantiation of the driver and reuse it until things are closed